_source/_posts/2019-09-24-how-to-create-a-symfony-app-with-authentication.md

```diff
@@ -0,0 +1,771 @@
+
+Modify the `HomeController` to implement the login and callback methods:
+
+`src/Controller/HomeController.php` (full version):
+
+```php
+<?php
+namespace App\Controller;
+
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+use App\Entity\User;
+use App\Repository\UserRepository;
+use App\Services\OktaApiService;
+
+class HomeController extends AbstractController
+{
+
+    private $session;
+    private $okta;
+    private $userRepository;
+
+    public function __construct(
+        SessionInterface $session,
+        OktaApiService $okta,
+        UserRepository $UserRepository)
+    {
+        $this->session = $session;
+        $this->okta = $okta;
+        $this->userRepository = $UserRepository;
+    }
+
+    /**
+    * @Route("/", name="home")
+    */
+    public function home()
+    {
+        return $this->render('home.html.twig');
+    }
+
+    /**
+    * @Route("/personal-home-page", name="personal")
+    */
+    public function personal()
+    {
+        return $this->render('personal.html.twig');
+    }
+
+    /**
+    * @Route("/login", name="login")
+    */
+    public function login()
+    {
+        return $this->redirect($this->okta->buildAuthorizeUrl());
+    }
+
+    /**
+    * @Route("/authorization-code/callback", name="callback")
+    */
+    public function callback()
+    {
+        $token = $this->okta->authorizeUser();
+
+        if (!$token) {
+            return $this->redirectToRoute('home');
+        }
+
+        $email = $token->email;
+        $user = $this->userRepository->findOneByEmail($email);
+
+        if (! $user) {
+            $user = new User();
+            $user->setEmail($email);
+            $user->setName($email);
+            $user->setToken($accessToken);
+            $em = $this->getDoctrine()->getManager();
+            $em->persist($user);
+            $em->flush();
+        }
+
+        // Manually authenticate the user
+        $token = new UsernamePasswordToken($user, null, 'main', $user->getRoles());
+        $this->get('security.token_storage')->setToken($token);
+        $this->get('session')->set('_security_main', serialize($token));
+
+        return $this->redirectToRoute('home');
+    }
+
+    /**
+    * @Route("/logout", name="logout")
+    */
+    public function logout()
+    {
+
+    }
+}
+```
+
+Clicking the login button now takes you to the Okta login form, or immediately logs you in if you're already logged into Okta. Once you're logged in, you can access the Personal Home Page link and you can also see your Okta email displayed as your login identifier.
+
+## Learn More About Symfony, Authentication, and Okta
+
+You can find the whole code example [on GitHub](https://github.com/oktadeveloper/okta-php-symfony-auth). 
```

Comment:
Repos are commonly suffixed with `-example`. Also, there's a typo in the README: 

> Symfony project usokg Okta.